packages/openapi3/src/schema-emitter.ts

```diff
@@ -0,0 +1,866 @@
+    type: Scalar | Model | ModelProperty | Union | Enum,
+    original: OpenAPI3Schema
+  ): ObjectBuilder<OpenAPI3Schema> {
+    const schema = new ObjectBuilder(original);
+    const program = this.emitter.getProgram();
+    const applyConstraint = (fn: (p: Program, t: Type) => any, key: keyof OpenAPI3Schema) => {
+      const value = fn(program, type);
+      if (value !== undefined) {
+        schema[key] = value;
+      }
+    };
+
+    applyConstraint(getMinLength, "minLength");
+    applyConstraint(getMaxLength, "maxLength");
+    applyConstraint(getMinValue, "minimum");
+    applyConstraint(getMaxValue, "maximum");
+    applyConstraint(getPattern, "pattern");
+    applyConstraint(getMinItems, "minItems");
+    applyConstraint(getMaxItems, "maxItems");
+
+    const minValueExclusive = getMinValueExclusive(program, type);
+    if (minValueExclusive !== undefined) {
+      schema.minimum = minValueExclusive;
+      schema.exclusiveMinimum = true;
+    }
+
+    const maxValueExclusive = getMaxValueExclusive(program, type);
+    if (maxValueExclusive !== undefined) {
+      schema.maximum = maxValueExclusive;
+      schema.exclusiveMaximum = true;
+    }
+
+    if (isSecret(program, type)) {
+      schema.format = "password";
+    }
+
+    // the stdlib applies a format of "url" but json schema wants "uri",
+    // so ignore this format if it's the built-in type.
+    if (!this.#isStdType(type) || type.name !== "url") {
+      applyConstraint(getFormat, "format");
+    }
+
+    applyConstraint(getDoc, "description");
+    applyConstraint(getSummary, "title");
+    applyConstraint(
+      (p: Program, t: Type) => (getDeprecated(p, t) !== undefined ? true : undefined),
+      "deprecated"
+    );
+
+    this.#attachExtensions(program, type, schema);
+
+    const values = getKnownValues(program, type as any);
+    if (values) {
+      return new ObjectBuilder({
+        oneOf: [schema, this.#enumSchema(values)],
+      });
+    }
+
+    return new ObjectBuilder(schema);
+  }
+
+  #inlineType(type: Type, schema: ObjectBuilder<any>) {
+    if (this.#options.includeXTypeSpecName !== "never") {
+      schema.set("x-typespec-name", getTypeName(type, this.#typeNameOptions()));
+    }
+    return schema;
+  }
+  #createDeclaration(type: Type, name: string, schema: ObjectBuilder<any>) {
+    const refUrl = getRef(this.emitter.getProgram(), type);
+    if (refUrl) {
+      return {
+        $ref: refUrl,
+      };
+    }
+
+    if (shouldInline(this.emitter.getProgram(), type)) {
+      return this.#inlineType(type, schema);
+    }
+
+    const usage = this.#visibilityUsage.getUsage(type);
+    const shouldAddSuffix = usage !== undefined && usage.size > 1;
+    const visibility = this.#getVisibilityContext();
+    const fullName =
+      name + (shouldAddSuffix ? getVisibilitySuffix(visibility, Visibility.Read) : "");
+
+    const decl = this.emitter.result.declaration(fullName, schema);
+    // TODO  is there a
+    checkDuplicateTypeName(
```

Comment:
Is there a better way to find duplicate declaration

Reply:
Not as yet, though could be added.